packages/lesswrong/lib/collections/bookmarks/newSchema.ts

```diff
@@ -0,0 +1,103 @@
+import { DEFAULT_CREATED_AT_FIELD, DEFAULT_ID_FIELD } from "@/lib/collections/helpers/sharedFieldConstants";
+import { userOwns } from "@/lib/vulcan-users/permissions";
+import gql from "graphql-tag";
+
+const ALLOWED_COLLECTION_NAMES = ["Posts", "Comments"];
+
+export const graphqlTypeDefs = gql`
+  union BookmarkableDocument = Post | Comment
+`;
+
+const schema = {
+  _id: DEFAULT_ID_FIELD,
+  createdAt: DEFAULT_CREATED_AT_FIELD,
+
+  documentId: {
+    database: {
+      type: "TEXT",
+      nullable: false,
+    },
+    graphql: {
+      outputType: "String",
+      inputType: "String!",
+      canRead: ['guests'],
+      canCreate: ['members'],
+      canUpdate: [userOwns],
+    },
+  },
+
+  collectionName: {
+    database: {
+      type: "TEXT",
+      nullable: false,
+    },
+    graphql: {
+      outputType: "String",
+      inputType: "String!",
+      canRead: ['guests'],
+      canCreate: ['members'],
+      validation: {
+        allowedValues: ALLOWED_COLLECTION_NAMES,
+      },
+    },
+  },
+
+  userId: {
+    database: {
+      type: "VARCHAR(27)",
+      foreignKey: "Users",
+      nullable: false,
+    },
+    graphql: {
+      outputType: "String",
+      canRead: ['guests'],
+      canCreate: ['members'],
+    },
+  },
+
+  post: {
+    graphql: {
+      outputType: 'Post',
+      canRead: ['guests'],
+      resolver: async (bookmark: DbBookmark, args: any, context: ResolverContext) => {
+        const { documentId, collectionName } = bookmark;
+        const { Posts } = context;
+        if (collectionName !== "Posts" || !documentId || !Posts) {
+            return null;
+        }
+        return await Posts.findOne({ _id: documentId });
```

Comment:
This should probably be using a loader... and in fact it should be using a `getWithLoader` to avoid an n+1 query, since I imagine you'll often be fetching a post for every single bookmark you're loading?

Reply:
why does that make "getWithLoader" necessary rather than just regular loader? There's no grouping

Reply:
It's about making a single database query for n records (since dataloaders will do the batching for you under the hood, assuming certain conditions are met), rather than making n database queries.